Support nn.Module forward hooks in torchdynamo

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94917
* #94951
* __->__ #92125
* #95248

Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Add failing tests for nn module hooks

Also add comments pointing to the places we probably
need to change dynamo to fix 2 categories of hooks issues.

[ghstack-poisoned]

wconstab committed:
Update on "Add failing tests for nn module hooks"


Also add comments pointing to the places we probably
need to change dynamo to fix 2 categories of hooks issues.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Add failing tests for nn module hooks"


Also add comments pointing to the places we probably
need to change dynamo to fix 2 categories of hooks issues.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Add failing tests for nn module hooks"


Also add comments pointing to the places we probably
need to change dynamo to fix 2 categories of hooks issues.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -211,35 +218,19 @@ def record_nn_module_stack():
-                        )
-                        args = [self] + args
+                fn = mod.__call__
+                fn_source = AttrSource(self.source, "__call__")
```
I remember we don't support nn Module only has ```__call__``` rather than ```forward``` function. Does it mean we will support it after this PR?

wconstab replied:
yea, I think this PR will fix modules that only have __call__, at least in the majority of cases.

I am going to write another comment that raises a related question, specifically for the top-level module that you apply torch.compile() to.

wconstab commented on torch/_dynamo/eval_frame.py:
```diff
@@ -79,7 +79,7 @@ def __getattr__(self, name):
 
     def forward(self, *args, **kwargs):
-        return self.dynamo_ctx(self._orig_mod.forward)(*args, **kwargs)
+        return self.dynamo_ctx(self._orig_mod.__call__)(*args, **kwargs)
```
This change looks innocuous enough.  But it's actually concerning me a bit.

We're asymmetrically mapping from `forward` of the outer 'OptimizedModule' object to `__call__` of the wrapped module.  This is probably not a good idea.

I would feel better if we could just change OptimizedModule implementation so it does not implement forward at all, and instead implements `__call__`.  Would this be OK though?  And does this even matter or is it just a nit?

voznesenskym replied:
Are you using __call__ here to force hitting the hooks?

wconstab replied:
right. referring to my notes in the Commit Message / PR Desc, This is the case (2) where hooks were applied to the outermost module m passed to compile(m), and would still be missed if we only support `__call__` inside NNModuleVariable.

And the reason changing `forward` to `__call__` helps with hooks is that torch NNModule implements __call__ to call hooks and then call forward, more or less.

ezyang replied:
> I would feel better if we could just change OptimizedModule implementation so it does not implement forward at all, and instead implements __call__. Would this be OK though? And does this even matter or is it just a nit?

Ditto, wondering why we don't just change line 81 to `__call__`. Also, it seems fine in principle to keep the forward stub around, going to inner forward.

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -496,6 +496,17 @@ def index_source(key):
                 ),
                 "apply",
             )
+        elif isinstance(value, types.MethodType):
```
nit, move this down to before UserDefinedObject?

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -496,6 +496,17 @@ def index_source(key):
+            # which doesn't support'CALL_FUNCTION'
+            return UserMethodVariable(
+                value.__func__,
+                VariableBuilder(self.tx, source=AttrSource(self.source, "__self__"))(
+                    value.__self__
+                ),
```
If this is `None` after re-entering the builder here, will that silently pass? Consider pulling this out, adding a comment for what we do, and checking a few basic assumptions.

voznesenskym commented on test/dynamo/test_modules.py:
```diff
@@ -1164,6 +1165,48 @@ def fn(x):
             )
         )
 
+    def test_hooks_outer(self):
```
Missing tests for:

1) Adding a hook, tracing, changing the hook, and ensuring a guard fails
2) What should export do with exporting a hook?
3) A hook that mutates state
4) A hook that graph breaks
5) A hook that invokes torch.compile

These might seem a little weird, but imo adding any new scope where we add support for user defined callables requires a lot of extra assurances.

wconstab replied:
yea, i will think more about those cases but yea in general agreed we need to test this more thoroughly, including the guarding aspects.

ezyang reviewed (approved):
this looks good, well done

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules, to route the call to __call__  instead of .forward(), since __call__ is the codepath that eager users hit and will dispatch to hooks correctly.
 1. inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 2. at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling module.forward() (which is a method) as a 
 UserFunctionVariable with an extra 'self' arg.  After tracing into module.__call__, there is no longer a special case for the eventual call into .forward, and it gets wrapped in a UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Fixes #91665

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented on test/dynamo/test_modules.py:
```diff
@@ -1164,6 +1165,106 @@ def fn(x):
+
+        # so we don't correctly remove our hook in this case. It just runs
+        """
+        Dynamo is aware of m._forward_hooks, but we aren't installing a guard on it
```
setting `config.guard_nn_modules=True` fixes this 

we can probably do something about this just  for hooks.. but.. slippery slope

jansel commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -211,35 +218,19 @@ def record_nn_module_stack():
                     "Must provide a valid source in order to inline, "
                     "since inlined function may have default args which must be guarded."
                 )
-                if is_lazy:
```
Does removing this trigger a recompile on lazy modules?

wconstab replied:
i didn't check that. (would any existing test cover that case?)

but i don't think so, because i should be making the behavior that used to apply to lazy modules now apply more widely.

jansel replied:
Might be worth running a simple example and checking how many compiles you get

wconstab replied:
I confirmed the behavior is the same on master and on my PR for a simple case I based off of the lazy module in `test/dynamo/test_modules.py` and added compile counts in more places.

My observation was that we do not capture 1 graph for the first try in either specialized or unspecialized case, but, we also do not recompile on a second run of either.  Does that address your concern?

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules, to route the call to __call__  instead of .forward(), since __call__ is the codepath that eager users hit and will dispatch to hooks correctly.
 1. inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 2. at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling module.forward() (which is a method) as a 
 UserFunctionVariable with an extra 'self' arg.  After tracing into module.__call__, there is no longer a special case for the eventual call into .forward, and it gets wrapped in a UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Fixes #91665

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules, to route the call to __call__  instead of .forward(), since __call__ is the codepath that eager users hit and will dispatch to hooks correctly.
 1. inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 2. at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling module.forward() (which is a method) as a 
 UserFunctionVariable with an extra 'self' arg.  After tracing into module.__call__, there is no longer a special case for the eventual call into .forward, and it gets wrapped in a UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Fixes #91665

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -540,6 +540,25 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.FUNCTION_MATCH),
             )
+        elif isinstance(value, types.MethodType):
```
This graph breaks like crazy because of the nn module assumptions here - I forked the PR post discussion to https://github.com/pytorch/pytorch/pull/95353/files - the primary change is adding an nnmodule check - because there are lots of methods on other things that *should* work going through this, but are probably not good candidates for the implicit inlining this forces.

wconstab replied:
ok i didn't see your comment until after i read your forked PR and cherry picked just this piece.  I wrote up a comment/TODO that aligns with what you said here. Thanks for your help figuring this out!

wconstab committed:
Update on "Support nn.Module forward hooks in torchdynamo"


Tweak dynamo behavior in 2 places when calling nn.Modules,
to route the call to __call__  instead of .forward(), since
__call__ is the codepath that eager users hit and will dispatch
to hooks correctly.
 (1) inside NNModuleVariable.call_function, which covers the common case
     of calling a module from code dynamo is already tracing
 (2) at the OptimizedModule layer, which is the entrypoint
     into a top-level nn.Module dynamo is about to compile

This exposes a new bug: NNModuleVariable used to special-case calling
module.forward() (which is a method) as a UserFunctionVariable with an extra
'self' arg.  After tracing into module.__call__, there is no longer a special
case for the eventual call into .forward, and it gets wrapped in a
UserDefinedObjectVariable following standard behavior of ._wrap().  UDOV can't be
called, so this broke some tests.

- Fix: add a new special case in _wrap() that treats methods as a UserDefinedMethod
  instead of UserDefinedObjectVariable.  Now, the forward method can be called.

Also, fix NNModuleVar.call_method routing forward back to __call__

[ghstack-poisoned]

voznesenskym reviewed (approved):
the 2-3% slowdown seems acceptable for something this important.

wconstab commented:
i'm landing this now with the regression.
- it's worthwhile to get the hook support _and_ clean up some tracing behavior as a better eng win
- we can probably optimize the perf back by clever changes to NNModule, but i don't want to block this on that

wconstab commented:
@pytorchbot merge

voznesenskym commented:
> * we can probably optimize the perf back by clever changes to NNModule, but i don't want to block this on that

We discussed this offline, big +1 :)

ngimel commented:
@wconstab are you planning on fixing the regression? Most users don't use hooks (except for maybe debugging), and forcing them to pay 2-3% for a feature they deliberately don't use (because hooks cost performance even in eager) is not great.

wconstab commented:
Yea, I plan to take a stab at this, or worst case i'll try to get someone else to help with it.  If we can't get all the perf back, though, I still think we should keep this landed, since it's essentially a correctness fix rather than a feature.

The PR is really not supporting hooks, its just making `__call__` really call instead of silently jumping to .forward.

awgu commented on torch/_dynamo/variables/builder.py:
```diff
@@ -540,6 +540,32 @@ def index_source(key):
+                self.tx, source=AttrSource(self.source, "__self__")
+            )(value.__self__)
+            assert self_obj and isinstance(
+                self_obj, VariableTracker
```
Curious: How can `VariaableBuilder.__call__()` return a `None` value or a non-`VariableTracker` instance such that this `assert` can be triggered?

wconstab replied:
the concern was a None rather than another type; i'm not sure if currently it can fail, but it was a defensive guard

XiaobingSuper commented:
@wconstab, we met a big performance regression on the CPU path(**>10%**), see https://github.com/pytorch/pytorch/issues/95684, https://github.com/pytorch/pytorch/issues/95682, and https://github.com/pytorch/pytorch/issues/95683, how about add a check whether the mod has a hook? if not, directly call `mod.forward`.

wconstab commented:
> @wconstab, we met a big performance regression on the CPU path(**>10%**), see #95684, #95682, and #95683, how about add a check whether the mod has a hook? if not, directly call `mod.forward`.

That's too bad, it's higher than the regressions I saw on most gpu models.

However, the proposal to add a check isn't so simple.  The check is actually already there in [nn.Module._call_impl](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1498).  The purpose of the check is to see if all the hook dicts are empty, and if so then directly call forward.  The problem is that for dynamo to correctly check at runtime that all those hook dicts are _still_ empty (compared to compile time), adds this performance penalty.

We may be able to fix it by changing torch.nn.Module.  The change would be something that notices any time users add a new hook, and directly notifies dynamo to invalidate the cache and recompile.  This would let dynamo skip the checks at runtime.  (It's not strictly safe, since people could probably work around the check).  @voznesenskym did you have a PR that partially attempted this last year, can you paste it?

ngimel commented:
Given that it's such a large regression, can we prioritize the fix @wconstab ?

psinger commented:
Hi @wconstab - I just saw that this PR should solve the issues of forward hooks with torch compile.
Is there any plan to merge this to master soon, or is it already merged? Thanks!

wconstab commented:
> Hi @wconstab - I just saw that this PR should solve the issues of forward hooks with torch compile.
Is there any plan to merge this to master soon, or is it already merged? Thanks!

@psinger This PR was merged a while ago, however the hook support added some perf overhead to dynamo for every model (esp those not using hooks), and some gaps have been pointed out in hook support (currently they don't work on builtin modules like nn.Conv - https://github.com/pytorch/pytorch/issues/96722).

I'm proposing a new stack to 
1) https://github.com/pytorch/pytorch/pull/98223 - disable hook support by default (to get perf back for non-hook users)
2) https://github.com/pytorch/pytorch/pull/98223 - fix the known gap for builtin modules, but in a somewhat unsatisfying way (adding graph breaks)

Please do let me know what your specific use case is, and whether it is addressed above or not.
